Bulk Email for course modes

efischer19 opened the pull request:
Hackathon XV project

efischer19 commented:
Sandbox for this is up at https://efischer19.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-send_email (may need to log in first, use the `staff@example.com` account).

FYI for potential review in a post-hackathon capacity:
- @sstack22 
- @catong 
- 2 of @nasthagiri @yro @jcdyer @sanfordstudent

efischer19 commented on lms/djangoapps/bulk_email/models.py:
```diff
@@ -115,6 +123,12 @@ def get_users(self, course_id, user_id=None):
+                enrollment_qset.filter(
+                    courseenrollment__mode=self.coursemodetarget.track.mode_slug
+                )
+            )
```
@schenedx Can you (or someone else on ECOM) confirm that this logic will correctly fetch all users enrolled in a given mode? I'm essentially grabbing all entries in the `CourseEnrollment` table [model link](https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/models.py#L969) that match the given `course_id` and `mode`, with `is_active` set to True.

efischer19 commented:
jenkins run python

efischer19 commented:
jenkins run python

efischer19 commented:
Huh, I've never seen a segfault during python execution before, that's new. FYI @edx/testeng - had a run of python tests start to error a bunch, then segfault and claim to have leaked file descriptors. Error log is here, I'm going to kick the job off to run again.

```
21:29:15 #3001 test_bad_user (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ok
21:29:15 #3002 test_course_id_string (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3003 test_course_name (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3004 test_get_and_set_credit_state (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3005 test_inactive_enrollment (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3006 test_invalid_user (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ok
21:29:16 #3007 test_no_profile_name (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ok
21:29:16 #3008 test_not_credit_course (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3009 test_remove_credit_requirement_status (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3010 test_remove_status_non_credit (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3011 test_set_status_non_credit (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3012 test_set_status_non_verified_enrollment_1_audit (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3013 test_set_status_non_verified_enrollment_2_honor (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3014 test_set_status_non_verified_enrollment_3_credit (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ERROR
21:29:16 #3015 test_user_not_enrolled (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ok
21:29:16 #3016 test_user_not_found (openedx.core.djangoapps.credit.tests.test_services.CreditServiceTests) ... ok
21:29:16 Fatal Python error: Segmentation fault
21:29:16 
21:29:16 Current thread 0x00007f845dd7e700 (most recent call first):
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py", line 316 in execute
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 62 in execute
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/db/backends/base/base.py", line 214 in _savepoint_rollback
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/db/backends/base/base.py", line 254 in savepoint_rollback
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/db/transaction.py", line 245 in __exit__
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/test/testcases.py", line 941 in _rollback_atomics
21:29:16   File "/home/jenkins/workspace/edx-platform-test-subset/common/djangoapps/util/testing.py", line 160 in _wrapper
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/test/testcases.py", line 998 in _fixture_teardown
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/test/testcases.py", line 853 in _post_teardown
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/test/testcases.py", line 189 in __call__
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/case.py", line 151 in runTest
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/case.py", line 133 in run
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/case.py", line 45 in __call__
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/suite.py", line 224 in run
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/suite.py", line 177 in __call__
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/suite.py", line 224 in run
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/suite.py", line 177 in __call__
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/core.py", line 62 in run
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/core.py", line 207 in runTests
21:29:16   File "/usr/lib/python2.7/unittest/main.py", line 95 in __init__
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/core.py", line 121 in __init__
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django_nose/runner.py", line 299 in run_suite
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django_nose/runner.py", line 352 in run_tests
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 90 in handle
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/base.py", line 445 in execute
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 74 in execute
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/base.py", line 394 in run_from_argv
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 30 in run_from_argv
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 346 in execute
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 354 in execute_from_command_line
21:29:16   File "./manage.py", line 116 in <module>
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/coverage/execfile.py", line 177 in run_python_file
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/coverage/cmdline.py", line 627 in do_run
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/coverage/cmdline.py", line 480 in command_line
21:29:16   File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/coverage/cmdline.py", line 753 in main
21:29:16   File "/home/jenkins/edx-venv/lib/python2.7/site-packages/coverage/__main__.py", line 8 in <module>
21:29:16   File "/usr/lib/python2.7/runpy.py", line 72 in _run_code
21:29:16   File "/usr/lib/python2.7/runpy.py", line 162 in _run_module_as_main
21:29:17 Segmentation fault (core dumped)
21:29:17 
21:29:17 Cleaning up after cms
21:29:17 ---> pavelib.utils.test.utils.clean_mongo
21:29:17 mongo localhost:27017 /home/jenkins/workspace/edx-platform-test-subset/scripts/delete-mongo-test-dbs.js
21:29:17 MongoDB shell version: 2.6.12
21:29:17 connecting to: localhost:27017/test
21:29:17 Dropping test db test_xcontent_2920
21:29:17 Dropping test db test_xmodule_2920
21:29:17 
21:29:17 Cleaning up after python tests
21:29:17 
21:29:17 
21:29:17 ================================================
21:29:17 Tests failed in the following suites:
21:29:17 * cms
21:29:17 
21:29:17 
21:29:27 Process leaked file descriptors. See http://wiki.jenkins-ci.org/display/JENKINS/Spawning+processes+from+build for more information
21:29:27 Build step 'Execute shell' marked build as failure
21:29:28 Archiving artifacts
21:29:30 [xUnit] [INFO] - Starting to record.
21:29:30 [xUnit] [INFO] - Processing JUnit
21:29:31 [xUnit] [INFO] - [JUnit] - No test report file(s) were found with the pattern '**/nosetests.xml,**/TEST-*.xml,reports/acceptance/*.xml,reports/quality.xml,reports/javascript/javascript_xunit.xml,reports/bok_choy/xunit.xml,reports/bok_choy/**/xunit.xml' relative to '/home/jenkins/workspace/edx-platform-test-subset' for the testing framework 'JUnit'.  Did you enter a pattern relative to the correct directory?  Did you generate the result report(s) for 'JUnit'?
21:29:31 [xUnit] [WARNING] - No test reports found for the metric 'JUnit' with the resolved pattern '**/nosetests.xml,**/TEST-*.xml,reports/acceptance/*.xml,reports/quality.xml,reports/javascript/javascript_xunit.xml,reports/bok_choy/xunit.xml,reports/bok_choy/**/xunit.xml'.
21:29:31 [xUnit] [INFO] - Skipping the metric tool processing.
21:29:31 [xUnit] [INFO] - There are errors when processing test results.
21:29:31 [xUnit] [INFO] - Skipping tests recording.
21:29:31 Finished: FAILURE
```

jzoldak commented:
@efischer19 I'm not seeing that segfault from other prs according to splunk (`index=testeng "Fatal Python error: Segmentation fault"`), though I do see it in 3 of your builds. So I'm thinking it is related. 
@cpennington do we have any documentation for how Eric would track this down?

cpennington commented:
That looks like the same one as we thought was fixed by upgrading sqlite.

Typically, the segfault is based on a specific ordering of tests. @efischer19, this feels hacky, but can you put your new test method before or after one of the other tests in the same class, and see if that changes things?

catong commented:
@efischer19 @sstack22 FYI will need to create a corresponding doc PR if this project is going to be available on edx.org.

efischer19 commented:
I've moved the new tests I wrote to different locations in their files and rebased, we'll see how that goes.

@catong Of course! Feel free to tag me on that whenever it gets made. This feature won't be going live until next week at the earliest, and I could easily hold it further if you need more time.

efischer19 commented:
@cpennington Still failing, any further suggestions of where to investigate next? This seems to be affecting only me, as JZ noted, but I haven't touched any of the tests in `openedx/core/djangoapps/credit`, where I'm reliably seeing the failure on jenkins. All of those tests run and pass locally.

efischer19 commented:
jenkins run python

efischer19 commented:
jenkins run python

efischer19 commented:
jenkins run all

cpennington commented:
It seems like there are other errors in the test suite, before the segfault. You could try (temporarily) commenting out the tests that segfault, and then fixing the other errors, and then uncommenting/skipping the segfaulting tests.

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
@nasthagiri @jcdyer @sanfordstudent @cpennington I've gotten python tests to pass here. I fixed an issue with tests in `cms/djangoapps/contentstore`, and that made the segfaults go away.

Does anyone have a better solution to "lms code is being imported in cms processes" than https://github.com/edx/edx-platform/pull/14118/commits/0e66acfb1648afe22decf26d3c0cea8c4cf95311?

nasthagiri commented:
👍

jcdyer commented:
:+1:

efischer19 commented:
Follow up ticket for the circular import: https://openedx.atlassian.net/browse/TNL-6298
